src/com/amazon/ion/impl/IonReaderBinaryIncremental.java

```diff
@@ -0,0 +1,1932 @@
+     */
+    private static class ContainerInfo {
+
+        /**
+         * The container's type.
+         */
+        private IonType type;
+
+        /**
+         * The byte position of the end of the container.
+         */
+        private int endPosition;
+    }
+
+    /**
+     * The standard {@link IonBufferConfiguration}. This will be used unless the user chooses custom settings.
+     */
+    private static final IonBufferConfiguration STANDARD_BUFFER_CONFIGURATION =
+        IonBufferConfiguration.Builder.standard().build();
+
+    // Constructs ContainerInfo instances.
+    private static final _Private_RecyclingStack.ElementFactory<ContainerInfo> CONTAINER_INFO_FACTORY =
+        new _Private_RecyclingStack.ElementFactory<ContainerInfo>() {
+
+        @Override
+        public ContainerInfo newElement() {
+            return new ContainerInfo();
+        }
+    };
+
+    // The Ion 1.0 system symbol table.
+    private static final List<String> SYSTEM_SYMBOLS_1_0 = Collections.unmodifiableList(new ArrayList<String>() {{
+        add(null);
+        add("$ion");
+        add("$ion_1_0");
+        add("$ion_symbol_table");
+        add("name");
+        add("version");
+        add("imports");
+        add("symbols");
+        add("max_id");
+        add("$ion_shared_symbol_table");
+    }});
+
+    // The size of the Ion 1.0 system symbol table.
+    private static final int SYSTEM_SYMBOLS_1_0_SIZE = SYSTEM_SYMBOLS_1_0.size();
+
+    // The system symbol table SID for the text "$ion_symbol_table".
+    private static final int ION_SYMBOL_TABLE_ID = 3;
+
+    // The system symbol table SID for the text "name".
+    private static final int NAME_ID = 4;
+
+    // The system symbol table SID for the text "version".
+    private static final int VERSION_ID = 5;
+
+    // The system symbol table SID for the text "imports".
+    private static final int IMPORTS_ID = 6;
+
+    // The system symbol table SID for the text "symbols".
+    private static final int SYMBOLS_ID = 7;
+
+    // The system symbol table SID for the text "max_id".
+    private static final int MAX_ID_ID = 8;
+
+    // The size of the reusable UTF-8 decoding buffer.
+    private static final int UTF8_BUFFER_SIZE_IN_BYTES = 4 * 1024;
+
+    // The final byte of the binary IVM.
+    private static final int IVM_FINAL_BYTE = 0xEA;
+
+    // Isolates the highest bit in a byte.
+    private static final int HIGHEST_BIT_BITMASK = 0x80;
+
+    // Isolates the lowest seven bits in a byte.
+    private static final int LOWER_SEVEN_BITS_BITMASK = 0x7F;
+
+    // Isolates the lowest six bits in a byte.
+    private static final int LOWER_SIX_BITS_BITMASK = 0x3F;
+
+    // The number of significant bits in each UInt byte.
+    private static final int VALUE_BITS_PER_UINT_BYTE = 8;
+
+    // The number of significant bits in each VarUInt byte.
+    private static final int VALUE_BITS_PER_VARUINT_BYTE = 7;
```

Comment:
> [...]got me to wondering why ResizingPipedInputStream.peek(int n) yields an int instead of a byte, given that it's peeking into a byte[].

Two reasons:
1. It's meant as an alternative to `@Override public int read()` from the `InputStream` interface, which `ResizingPipedInputStream` implements. For consistency, returning an `int` from `peek` felt right. This is admittedly not a great reason.
2. Most places `peek` is used, its return value would be upgraded to an `int` anyway. E.g. indexing into the TYPE_IDS array, being read into [Var][U]Int values, etc. There are [some indications](https://stackoverflow.com/questions/14531235/in-java-is-it-more-efficient-to-use-byte-or-short-instead-of-int-and-float-inst) that doing math with `byte` may actually be slower than with `int`.